### DISCUSSION TITLE
Modelando Domínios Ricos - Fail Fast Validations

### DISCUSSION THREAD

**Discussion by ddcsilva** [balta-io/forum#244] [Tira Dúvidas] [answered]
Opa! Curtindo demais o curso sobre domínios ricos!

Tenho uma dúvida, talvez tenha sido falado, mas não entendi.

Na aula de Fail Fast Validations, temos um command por exemplo pra criar pagamento de cartão de crédito.  Implementamos a interface o iCommand que tem um método de validar e herdamos de Notifiable.  Mas os atributos ali são copy paste da implementação dos value objects, não daria para usar os value objects direto?

Ali no método validate(), no vídeo é colocado a mesma validação que adicionei no meu contract. Exemplificando. Nesse Command que pelo que entendi serve como um DTO, eu devo evitar tipos complexos e mandar somente valores primitivos, por causa que são dados que o front terá que servir, ou não tem nada a ver?

Eu criei uma classe de Contract para meu NomeCompleto, por exemplo, ai eu chamaria ele no meu validate, estaria incorreto?

```
public class CriarNomeCompletoContract : Contract<NomeCompleto>
    {
        public CriarNomeCompletoContract(NomeCompleto nomeCompleto)
        {
            Requires()
                .IsGreaterOrEqualsThan(nomeCompleto.Nome, 3, "NomeCompleto.Nome", "Nome deve conter pelo menos 3 caracteres")
                .IsGreaterOrEqualsThan(nomeCompleto.Sobrenome, 3, "NomeCompleto.Sobrenome", "Sobrenome deve conter pelo menos 3 caracteres")
                .IsLowerOrEqualsThan(nomeCompleto.Nome, 40, "NomeCompleto.Nome", "Nome deve conter até 40 caracteres");
        }
    }
```

PS: Misturei Português e Inglês só pra associar alguns conceitos, mas na prática não faço isso não :rofl:

**Comment by andrebaltieri** [comment, marked as answer]
Bom dia, @ddcsilva, como vai?

"não daria para usar os value objects direto?"
- Daria sim, é uma forma de fazer. Eu particularmente prefiro deixar separado. Se precisar altero só o Command e não afeta nada meu VO.

"eu devo evitar tipos complexos e mandar somente valores primitivos, por causa que são dados que o front terá que servir, ou não tem nada a ver?"
- Exatamente... a ideia é simplificar para o front, enviando apenas estruturas mais simples.
